Source/Mosa.Utility.BootImage/Limine.cs

```diff
@@ -0,0 +1,1618 @@
+	0x69, 0x41, 0x71, 0x59, 0x36, 0x39, 0xb6, 0xdc, 0xa3, 0x50, 0x08, 0x1e, 0xb7, 0x3a, 0x8b, 0x44,
+	0xee, 0xe7, 0xad, 0xc8, 0x0d, 0x11, 0x93, 0x90, 0xcb, 0x55, 0xb5, 0xe2, 0x10, 0x2e, 0x2f, 0x10,
+	0x23, 0xe5, 0x6f, 0x43, 0x06, 0x2c, 0xa3, 0xc5, 0x62, 0xa9, 0x80, 0x33, 0x7f, 0x88, 0x44, 0xd0,
+	0x50, 0x71, 0x73, 0xe8, 0xbd, 0xac, 0x04, 0x96, 0x0b, 0xdc, 0x62, 0x78, 0x87, 0x0a, 0x25, 0x25,
+	0xb0, 0x64, 0x2b, 0xa2, 0x77, 0x5c, 0x00, 0x95, 0x72, 0x15, 0x1f, 0x52, 0x5b, 0x54, 0x68, 0xaf,
+	0xa9, 0xaf, 0xa4, 0xdc, 0xce, 0x84, 0xb3, 0xc1, 0xcd, 0x1c, 0x70, 0x46, 0x7d, 0xcd, 0x01, 0x11,
+	0x82, 0xe4, 0xe2, 0x33, 0xe1, 0x5a, 0xf0, 0x2e, 0xab, 0xef, 0x2b, 0x41, 0x4e, 0x5f, 0x3e, 0x83,
+	0x38, 0x02, 0x25, 0x25, 0x62, 0x71, 0x11, 0x37, 0xe4, 0x6c, 0x44, 0x5c, 0xca, 0x7a, 0x17, 0x36,
+	0xdc, 0x4e, 0xcc, 0x04, 0x21, 0x5b, 0xc6, 0x79, 0xcf, 0x61, 0x71, 0x11, 0xaa, 0xb6, 0xd2, 0x7e,
+	0x62, 0xb6, 0x27, 0x5c, 0x00, 0x40, 0xfb, 0xe5, 0x7f, 0x1e, 0xce, 0x2f, 0xe8, 0x97, 0xfb, 0xc2,
+	0x47, 0x6a, 0xae, 0xd4, 0x3f, 0x7b, 0xea, 0xc7, 0x6a, 0xf6, 0xda, 0x7e, 0xb9, 0x7f, 0x82, 0xdc,
+	0xd5, 0x2a, 0x5f, 0x1e, 0x3a, 0x1f, 0xc6, 0x8e, 0x5b, 0x7a, 0x1e, 0x09, 0x00, 0x47, 0x7d, 0x8a,
+	0x3a, 0x21, 0xc7, 0x7d, 0x0a, 0x42, 0x82, 0x8b, 0x65, 0x3c, 0x74, 0x87, 0x17, 0xc4, 0xff, 0x1f,
+	0x06, 0x64, 0x6c, 0xb7, 0x50, 0x7e, 0x00, 0x00
+};
+
+		enum CacheState { Clean, Dirty }
+
+		private static FileStream Stream;
+		private static byte[] Cache;
+		private static ulong CachedBlock;
+		private static CacheState State;
+
+		private const int BlockSize = 512;
+
+		public static unsafe void Deploy(string path)
+		{
+			Stream = File.Open(path, FileMode.Open);
+			Cache = new byte[BlockSize];
+			CachedBlock = 0;
+			State = CacheState.Clean;
+
+			Stream.Read(Cache);
+
+			// Skip detecting if the device has a valid partition table, because we know it does
+
+			var stage2Size = bootloader_img.Length - 512;
+			var stage2Sectors = DivRoundUp(stage2Size, 512);
+			var stage2SizeB = (ushort)(stage2Sectors / 2 * 512);
+			var stage2SizeA = (ushort)(stage2SizeB + ((stage2Sectors % 2 != 0) ? 512 : 0));
+
+			// Default split of stage 2 for MBR (consecutive in post MBR gap)
+			ulong stage2LocA = 512;
+			var stage2LocB = stage2LocA + stage2SizeA;
+
+			var origMBR = new byte[70];
+			var timestamp = new byte[6];
+
+			// Save original timestamp
+			Read(timestamp, 218, 6);
+
+			// Save the original partition table of the device
+			Read(origMBR, 440, 70);
+
+			// Write the bootsector from the bootloader to the device
+			Write(bootloader_img, 0, 512);
+
+			// Write the rest of stage 2 to the device
+			Write(bootloader_img.Skip(512).ToArray(), stage2LocA, stage2SizeA);
+			Write(bootloader_img.Skip(512 + stage2SizeA).ToArray(), stage2LocB, (ulong)(stage2Size - stage2SizeA));
+
+			// Hardcode in the bootsector the location of stage 2 halves
+			// It seems like we need to use pointers here
+			stage2SizeA = EndSwap(stage2SizeA);
+			WritePtr(&stage2SizeA, 0x1a4, sizeof(ushort));
```

Comment:
No unsafe code! 

See Mosa.Utility.BootImage\Syslinux.cs and Generator.cs as examples. 

Note: We are not supporting big endian (all the code to do that was removed a few years ago).